### ISSUE TITLE
Parsing error with polymorphic anonymous functions 

### ISSUE THREAD

**Issue by bodhi** [babel/babel-eslint#266]
Sorry if this is the wrong repo to post this on!

With this file:

```
/* @flow */
function a<U>(b: U): U {
    return b
}

const a2 = <U>(b: U): U => {
    return b
}
```

I get this error running ESLint with Babel-eslint:

```
a.js
  7:5  error  Parsing error: Unexpected token
```

Which is the `return` in the body of `a2`. With an implicit return, I get a different error:

```
const a3 = <U>(b: U): U => b

=>  xx:15  error  Parsing error: Unterminated JSX contents
```

Flow has no problems with any of the declarations.

**Comment by bodhi** [comment]
Oops, forgot to say I saw this with v5.0.0 and v6.0.0-beta.1.

**Comment by hzoo** [comment]
If it's valid in flow, then it will need to be supported in babylon (babel's parser)

**Comment by danez** [comment]
This is fixed in babylon 6.8.2.
